ur_robot_driver/src/ur/dashboard_client.cpp

```diff
@@ -0,0 +1,347 @@
+  }
+  return true;
+}
+
+bool DashboardClient::brakeRelease()
+{
+  std::string answer = sendAndReceive("brake release\n");
+  if (answer != "Brake releasing")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::clearOperationalMode()
+{
+  std::string answer = sendAndReceive("clear operational mode\n");
+  if (answer != "operational mode is no longer controlled by Dashboard Server")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::closePopup()
+{
+  std::string answer = sendAndReceive("close popup\n");
+  if (answer != "closing popup")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::closeSafetyPopup()
+{
+  std::string answer = sendAndReceive("close safety popup\n");
+  if (answer != "closing safety popup")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::loadInstallation(const std::string& installation_name)
+{
+  std::string answer = sendAndReceive("load installation " + installation_name + "\n");
+  if (answer != "Loading installation: " + installation_name)
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::loadProgram(const std::string& program_name)
+{
+  std::string answer = sendAndReceive("load " + program_name + "\n");
+  if (answer != "Loading program: " + program_name + "\n")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::pause()
+{
+  std::string answer = sendAndReceive("pause");
+  if (answer != "Pausing program\n")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::play()
+{
+  std::string answer = sendAndReceive("play\n");
+  if (answer != "Starting program\n")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::popup(const std::string& text)
+{
+  std::string answer = sendAndReceive("popup " + text + "\n");
+  if (answer != "showing popup\n")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::powerOff()
+{
+  std::string answer = sendAndReceive("power off\n");
+  if (answer != "Powering off\n")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::powerOn()
+{
+  std::string answer = sendAndReceive("power on\n");
+  if (answer != "Powering on\n")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::quit()
+{
+  std::string answer = sendAndReceive("quit\n");
+  if (answer != "Disconnected\n")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::restartSafety()
+{
+  std::string answer = sendAndReceive("restart safety\n");
+  if (answer != "Restarting safety\n")
+  {
+    return false;
+  }
+  return true;
+}
+
+bool DashboardClient::setOperationalMode(const OperationalMode mode)
+{
+  std::string mode_string;
+  switch (mode)
+  {
+    case OperationalMode::MANUAL:
+      mode_string = "manual";
+      break;
+    case OperationalMode::AUTOMATIC:
+      mode_string = "automatic";
+      break;
+    default:
+      throw std::runtime_error("Enum value not known.");
```

Comment:
Consider `throw`ing `std::invalid_argument`?

This also appears to be the only place where an exception is thrown, instead of returning `false`. Is that intentional or a deviation?

Reply:
This also got removed in the latest refactoring. This inconsistency was one of the reasons to refactor this.